### PR TITLE
SAK-39953: resources > add ability to define custom copyright options

### PR DESCRIPTION
**Pull request by bjones86**
https://jira.sakaiproject.org/browse/SAK-39953

This PR introduces the following new features:

* ability for an institution to define and use an arbitrary number of custom copyright options
** enabled/disabled via sakai.property "copyright.useCustom", which defaults to false
** custom copyright options are defined via /content/content-copyright/impl/src/bundle/org/sakaiproject/content/copyright/copyright.properties
* ability to force users to explicitly select copyright status for files
** enabled/disabled via sakai.property "copyright.requireChoice", which defaults to false
** works with both default and custom copyright options
** if set to true, inputs a "Please select a copyright status" entry in the copyright drop down (always first element in the list)
** client and server side validation to reject saving the resource if the drop down was not set to a value other than the "Please select..."
* adds the copyright drop down form field to the upload resources form (accessed from both the Resources tool and the Drop Box tool)
** one drop down, setting applies to all files uploaded
** previously was only present on the 'Edit Details' form

The linked PR also includes refactoring existing copyright code, removing old and unused copyright sakai.properties, and improved documentation of the current state of valid copyright sakai.properties in default.sakai.properties.

The following sakai.properties have been added:

```
# SAK-39953 - use custom copyright options rather than the predefined list
# NOTE: define your custom copyright options in /content/content-copyright/impl/src/bundle/org/sakaiproject/content/copyright/copyright.properties
# NOTE: examples are defined in the above file, with the keys 'custom.copyright.#'; define as many or as few as needed by your institution
# DEFAULT: false
# copyright.useCustom=true

# SAK-39953 - require users to explicity define the copyright status for uploaded files in the Resources tool
# NOTE: does not require any other copyright sakai.properties to be set (works with defaults, copyright.types and custom copyright)
# DEFAULT: false
# copyright.requireChoice=true
```

The following sakai.properties have been removed:

```
## NOTE: these copyright properties may not be valid anymore - 8 Sept 2011
# default.copyright=Copyright status is not yet determined.
# default.copyright.alert=true
# fairuse.url=http://fairuse.stanford.edu
# newcopyrightinput=true
```

* `default.copyright` was made obsolete via KNL-1399 (and re-implemented in this PR)
* `default.copyright.alert` was some legacy property, not really sure what it once did
* `fairuse.url` from sakai.properties was not being used; the functionality of overriding the default value was moved into message bundles (`/content/content-tool/tool/src/bundle/right.properties`) some time ago, so I provided some more documentation there about what it does and how to customize it properly
* `newcopyrightinput` was another legacy property not being used; not sure what it once did
** for all of the above: removed some lingering code in `/content/` project that was still referencing these and sticking them in the `STATE` but never actually using it

### CODE REVIEW

File: content/content-tool/tool/src/java/org/sakaiproject/content/tool/CopyrightDelegate.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+ * Used to centralize any copyright related tasks
+ * See SAK-39953
+ *
+ * @author bbailla2
+ */
+public class CopyrightDelegate
+{
+	private String copyrightInfo;	// Copyright info takes its value from the dropdown (eg. "I hold copyright")
+	private String copyrightStatus;	// Copyright status takes its value from the text area (ie. when "Use copyright below." is selected)
+	private boolean copyrightAlert;	// copyrightAlert is... ?
```
Comment (plukasew) [inline content/content-tool/tool/src/java/org/sakaiproject/content/tool/CopyrightDelegate.java]: copyrightAlert is... ?

File: content/content-tool/tool/src/java/org/sakaiproject/content/tool/CopyrightDelegate.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	 * Captures copyright properties from the ParameterParser
+	 * @param params the ParameterParser to pull the copyright properties from
+	 * @param li the copyright properties will be set on li if it exists, otherwise they are stored in this class' members
+	 */
+	public void captureCopyright(ParameterParser params, ListItem li)
+	{
+		String copyright = StringUtils.trimToNull(params.getString("copyright"));
+		if (copyright == null)
+		{
+			// do nothing -- there must be no copyrightDialog
```
Comment (plukasew) [inline content/content-tool/tool/src/java/org/sakaiproject/content/tool/CopyrightDelegate.java]: Could just return early here and avoid the rightward drift for the rest of the method

File: content/content-tool/tool/src/java/org/sakaiproject/content/tool/CopyrightDelegate.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+		}
+		else
+		{
+			String newCopyright = StringUtils.trimToNull(params.getString("newcopyright"));
+			boolean crAlert = params.getBoolean("copyrightAlert");
+			if (li == null)
+			{
+				this.copyrightInfo = copyright;
```
Comment (plukasew) [inline content/content-tool/tool/src/java/org/sakaiproject/content/tool/CopyrightDelegate.java]: Unnecessary this

File: content/content-tool/tool/src/webapp/vm/content/sakai_resources_properties.vm [hunk 4]
```diff
@@ -104,5 +109,29 @@
 #set($count = $count + 1)
 #end
 #end
+
+	// Client side copyright check
+	var submitButton = document.querySelector("#finish_button");
```
Comment (plukasew) [inline content/content-tool/tool/src/webapp/vm/content/sakai_resources_properties.vm]: getElementById can be used directly here

File: content/content-tool/tool/src/webapp/vm/resources/sakai_create_uploads.vm [hunk 5]
```diff
@@ -27,6 +29,35 @@
...
+                        <option value="$copyright.getType()" #if($copyright_defaultType == $copyright.getType())selected="selected"#end #if ($!copyright.getLicenseUrl())title="${copyright.getLicenseUrl()}"#end >
+                            $copyright.getText()
+                        </option>
+                #end
+                    </select>
+                    <small>
+                        (<a href="#" onclick="openMoreInfoCopyrightWindow('copyright');" title="[$tlang.getString('cright.newwin')]"><span class="skip"> [$tlang.getString('cright.newwin')]</span>$tlang.getString('cright.fairuse')</a>)
+                    </small>
+                </div>
+                <div id="copyright_status"#if($model.copyrightStatus == $USE_THIS_COPYRIGHT) style="display:block;"#else style="display:none;"#end>
```
Comment (plukasew) [inline content/content-tool/tool/src/webapp/vm/resources/sakai_create_uploads.vm]: inline styles

File: content/content-tool/tool/src/webapp/vm/resources/sakai_create_uploads.vm [hunk 6]
```diff
@@ -216,9 +253,25 @@ Dropzone.options.fileUploader = {
...
 
+        // Client side copyright check
         submitButton.addEventListener("click", function() {
-            submitButton.disabled=true;
-            dz.processQueue(); // Tell Dropzone to process all queued files.
+            var requireChoice = $copyright_requireChoice;
+            var copyrightSelect = document.getElementById('copyright');
+
+            // 'Please select...' is always the first element in the drop down, but only enforce this if configured via sakai.properties
+            if (requireChoice && copyrightSelect !== null && copyrightSelect.selectedIndex < 1)
```
Comment (plukasew) [inline content/content-tool/tool/src/webapp/vm/resources/sakai_create_uploads.vm]: is this chunk of script code duplication?

File: content/content-tool/tool/src/webapp/vm/resources/sakai_properties.vm [hunk 7]
```diff
@@ -207,23 +207,28 @@
 					<option value="$model.copyrightStatus" selected="selected">${model.copyrightStatus}</option>
 				#end
 			</select>
+			<small>
+				(<a href="#" onclick="openMoreInfoCopyrightWindow('copyright');" title="[$tlang.getString('cright.newwin')]"><span class="skip"> [$tlang.getString('cright.newwin')]</span>$tlang.getString('cright.fairuse')</a>)
+			</small>
 		</p>
-		<div id="copyright_status${DOT}$i"#if($model.copyrightStatus == $USE_THIS_COPYRIGHT) class="form-group" style="display:block;"#else style="display:none;"#end>
+		<div id="copyright_status"#if($model.copyrightStatus == $USE_THIS_COPYRIGHT) class="form-group" style="display:block;"#else style="display:none;"#end>
```
Comment (plukasew) [inline content/content-tool/tool/src/webapp/vm/resources/sakai_properties.vm]: inline styles

File: content/content-tool/tool/src/webapp/vm/resources/sakai_properties.vm [hunk 8]
```diff
@@ -207,23 +207,28 @@
...
+				<label for="newcopyright">
 					$tlang.getString("cright.info")
 				</label>
-				<textarea class="form-control" name="newcopyright${DOT}$i" id="newcopyright${DOT}$i" rows="2" wrap="virtual" >$validator.escapeHtmlTextarea("$!model.copyrightInfo")</textarea>
+				<textarea class="form-control" name="newcopyright" id="newcopyright" rows="2" wrap="virtual" >$validator.escapeHtmlTextarea("$!model.copyrightInfo")</textarea>
 			</p>
 		</div>
 		<div class="form-group">
-			<label for="copyrightAlert${DOT}$i" style="width:15.5em;display:block">
+			<label for="copyrightAlert" style="width:15.5em;display:block">
```
Comment (plukasew) [inline content/content-tool/tool/src/webapp/vm/resources/sakai_properties.vm]: inline styles

File: library/src/webapp/content/copyright/custom.copyright.1.html [hunk 9]
```diff
@@ -0,0 +1,40 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
+<html>
+	<head>
+		<meta http-equiv="Content-type" content="text/html;charset=UTF-8">
+		<meta http-equiv="Refresh" content="0:URL=https://fairuse.stanford.edu/overview/faqs/copyright-ownership/#what_are_the_exceptions_to_the_rule_that_the_creator_of_a_work_owns_the_copyright">
```
Comment (plukasew) [inline library/src/webapp/content/copyright/custom.copyright.1.html]: Should there be some comments in these Stanford files to make it clear they are just examples?

File: content/content-tool/tool/src/webapp/vm/resources/sakai_create_uploads.vm (hunk 6 above)
Comment (bjones86) [reply content/content-tool/tool/src/webapp/vm/resources/sakai_create_uploads.vm]: Yes, but pieces of the `else` statement are unique. I'll extract the common parts to a function in `sakai_properties.vm` (which both `sakai_resources_properties.vm` and `sakai_create_uploads.vm` consume).

File: library/src/webapp/content/copyright/custom.copyright.1.html (hunk 9 above)
Comment (bjones86) [reply library/src/webapp/content/copyright/custom.copyright.1.html]: I've provided some comments at the top of the files.

**Comment by bjones86** [comment]
All comments from Paul's review have been addressed.

**Comment by bjones86** [comment]
The remaining Codacy flags are either false positives (2), or issues with existing code which is not being touched (10).

**Comment by jonespm** [comment]
Merging this one, @bjones86 is going to add a property change so this can be tested on experimental.

**Pull request merged**